bnd.bnd

```diff
@@ -0,0 +1,10 @@
+Bundle-SymbolicName: org.jspecify.jspecify
```

Comment:
What is the usage of `SymbolicName`? Is it common to use some ID that isn't used for anything else?

Reply:
https://cwiki.apache.org/confluence/display/SMX4/Creating+a+Bundle+Using+Maven seems to think that it makes sense to match this to the Maven artifact ID, but I think that's the unqualified artifact ID (which would be something like "guava" or "jspecify"), so that strikes me as iffy, like maybe they wanted to show a non-default behavior as a demonstration and they made a suboptimal choice? Guava [doesn't do that](https://github.com/google/guava/blob/4a2d3025250664ff248f5e82373a13f70ce98faa/guava/pom.xml#L73), so it ends up with `Bundle-SymbolicName: com.google.guava`.

At this particular moment, [this Google cache of a Red Hat doc](https://webcache.googleusercontent.com/search?q=cache%3Ahttps%3A%2F%2Fdocs.redhat.com%2Fen%2Fdocumentation%2Fred_hat_jboss_fuse%2F6.0%2Fhtml%2Fdeploying_into_the_container%2Fbestpractices-buildbundles&oq=cache%3Ahttps%3A%2F%2Fdocs.redhat.com%2Fen%2Fdocumentation%2Fred_hat_jboss_fuse%2F6.0%2Fhtml%2Fdeploying_into_the_container%2Fbestpractices-buildbundles&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIGCAEQRRg60gEIMTAwN2owajeoAgCwAgA&sourceid=chrome&ie=UTF-8) recommends either `${project.artifactId}` or `${project.groupId}.${project.artifactId}`. This PR implements the latter, and I guess I feel better about that, given what I said above?

Reply:
bnd should auto fill in many of these. Caffeine doesn't specify it explicitly and its populated when I look at the jar's `META-INF/MANIFEST.MF`. I include a PAX-Exam test to validate that the jar works in under OSGi.

<details><summary>Caffeine's bundle</summary>

```ini
Manifest-Version: 1.0
Automatic-Module-Name: com.github.benmanes.caffeine
Build-Jdk-Spec: 11
Bundle-Description: A high performance caching library
Bundle-License: https://www.apache.org/licenses/LICENSE-2.0
Bundle-ManifestVersion: 2
Bundle-Name: com.github.ben-manes.caffeine
Bundle-SymbolicName: com.github.ben-manes.caffeine
Bundle-Version: 3.1.9.SNAPSHOT
Export-Package: com.github.benmanes.caffeine.cache;uses:="com.github.b
 enmanes.caffeine.cache.stats";version="3.1.9",com.github.benmanes.caf
 feine.cache.stats;uses:="com.github.benmanes.caffeine.cache";version=
 "3.1.9",com.github.benmanes.caffeine;version="3.1.9"
Implementation-Title: A high performance caching library
Implementation-Version: 3.1.9-SNAPSHOT
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=11))"
```
</details>

Reply:
> bnd should auto fill in many of these

This is true ([see](https://bnd.bndtools.org/heads/bundle_symbolicname.html)). 

**BUT**, since it is a "unique" identifier which other OSGi bundles will use to specify a dependency to - so  it should live forever. I would not rely on this to be auto-generated. 

From [OSGi Docs](https://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#framework.module.bsn):

> The Bundle-SymbolicName manifest header is a mandatory header. The bundle symbolic name and bundle version identify a unique bundle.

We had an issue in the past where we had  to rename a bundle folder (e.g. the project folder from `fooA.bar` to `fooB.bar` ... silly example). This changed the bsn too to `fooB.bar` too. This broke all other OSGi  bundles relying on `fooA.bar`. We could fix it by hardcoding the bsn in the `bnd.bnd` file, so it is independent of any folder naming. 

If this will not happen (who knows), then fine, but I personally would recommend to hard code it, so that you do not rely on auto-generation.

Reply:
> `${project.groupId}.${project.artifactId}`. This PR implements the latter,

Yes. But I admit I also searched around a bit and look how other bundles are doing it. 
For example: https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core/2.17.2

```
<!-- https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core -->
<dependency>
    <groupId>com.fasterxml.jackson.core</groupId>
    <artifactId>jackson-core</artifactId>
    <version>2.17.2</version>
</dependency>
```

**MANIFEST.MF (excerpt)**:

```
Bundle-License: https://www.apache.org/licenses/LICENSE-2.0.txt
Bundle-SymbolicName: com.fasterxml.jackson.core.jackson-core
Implementation-Vendor-Id: com.fasterxml.jackson.core
Specification-Title: Jackson-core
Bundle-DocURL: https://github.com/FasterXML/jackson-core
Bundle-Name: Jackson-core
Multi-Release: true
Build-Jdk-Spec: 1.8
Bundle-Description: Core Jackson processing abstractions (aka Streamin
 g API), implementation for JSON
Implementation-Title: Jackson-core
Implementation-Version: 2.16.1
Bundle-ManifestVersion: 2
Specification-Vendor: FasterXML
Bundle-Vendor: FasterXML
```

Reply:
Ooo, ouch, yes, the docs appear to be correct that the default is the directory name. I verified that it doesn't even respect [the project name](https://github.com/jspecify/jspecify/blob/f0af7de6f72d01e3a8f3ffc9f528540de770f987/settings.gradle#L21), so it's 100% what you name the local directory on your system.

If there were a way to automatically derive the SymbolicName from the Maven artifact ID or something (as in the Maven plugin?), then I could see setting us up for that. But being explicit seems fine to me.

I wonder how Caffeine's works.... But I don't wonder enough to try to figure it out :)

Reply:
The rest of the `Bundle-` header look like they also override generic defaults or add otherwise absent values, so those look good. And I see that `-noextraheaders` saves us from `Created-By` and `Tool` headers with specific versions in them.

Reply:
oh, caffeine works because I completely missed that I hard coded the value like @chrisrueger suggested. It was just in the project specific build file,

```kts
tasks.jar {
  from(sourceSets["main"].output + sourceSets["codeGen"].output)
  dependsOn(compileCodeGenJava)
  bundle.bnd(mapOf(
    "Bundle-SymbolicName" to "com.github.ben-manes.caffeine",
    "Import-Package" to "",
    "Export-Package" to listOf(
      "com.github.benmanes.caffeine",
      "com.github.benmanes.caffeine.cache",
      "com.github.benmanes.caffeine.cache.stats").joinToString(","),
    "Automatic-Module-Name" to "com.github.benmanes.caffeine"))
}
```